File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidHelper(root, null, null);
    }
    
    private boolean isValidHelper(TreeNode root, TreeNode minVal, TreeNode maxVal) {
        if (root == null) {
            return true;
        }
        
        if ((minVal != null && root.val <= minVal.val) || (maxVal != null && root.val >= maxVal.val)) {
            return false;
        }
        
        return isValidHelper(root.left, minVal, root) && isValidHelper(root.right, root, maxVal);
    }
}

File: LeetCode/0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValidHelper(root, float('-inf'), float('inf'))
    
    def isValidHelper(self, root: Optional[TreeNode], minVal, maxVal) -> bool:
        if root is None:
            return True
        if not (minVal < root.val < maxVal):
            return False
        return (self.isValidHelper(root.left, minVal, root.val) and 
                self.isValidHelper(root.right, root.val, maxVal))